File: api/routes/alerts.py
"""Price alert CRUD — Pro-only feature.

Users can set up to 10 alerts per symbol/condition/target combination.
Alerts are checked on each cron refresh and trigger an email if price
crosses the target. Once triggered, the alert stays in the list (marked
triggered=TRUE) so users can see their history.
"""
from __future__ import annotations

import logging
from typing import Any, Dict, List

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, field_validator

from ..auth import get_current_user
from ..db import get_conn
from ..models import User
from ..tier import PRO_ALERT_LIMIT

router = APIRouter(prefix="/api/alerts", tags=["alerts"])
log = logging.getLogger(__name__)


class AlertCreate(BaseModel):
    symbol: str
    condition: str  # "above" | "below"
    target: float

    @field_validator("symbol")
    @classmethod
    def normalize_symbol(cls, v: str) -> str:
        return v.strip().upper()

    @field_validator("condition")
    @classmethod
    def validate_condition(cls, v: str) -> str:
        if v not in ("above", "below"):
            raise ValueError("condition must be 'above' or 'below'")
        return v

    @field_validator("target")
    @classmethod
    def validate_target(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("target price must be positive")
        return round(v, 4)


def _list_alerts_for_user(user_id: str) -> List[Dict[str, Any]]:
    with get_conn() as c:
        with c.cursor() as cur:
            cur.execute(
                """SELECT id, symbol, condition, target, triggered, created_at
                   FROM price_alerts
                   WHERE user_id = %s
                   ORDER BY created_at DESC""",
                [user_id],
            )
            rows = cur.fetchall()
    return [
        {
            "id": r[0],
            "symbol": r[1],
            "condition": r[2],
            "target": float(r[3]),
            "triggered": r[4],
            "created_at": r[5].isoformat() if r[5] else None,
        }
        for r in rows
    ]


@router.get("")
def list_alerts(user: User = Depends(get_current_user)) -> List[Dict[str, Any]]:
    """Return all price alerts for the authenticated Pro user."""
    return _list_alerts_for_user(user.id)


@router.post("", status_code=status.HTTP_201_CREATED)
def create_alert(
    payload: AlertCreate,
    user: User = Depends(get_current_user),
) -> Dict[str, Any]:
    """Create a price alert. Max 10 active alerts per Pro user."""
    existing = _list_alerts_for_user(user.id)
    active_count = sum(1 for a in existing if not a["triggered"])
    if active_count >= PRO_ALERT_LIMIT:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail=f"Alert limit reached ({PRO_ALERT_LIMIT} active alerts max). Delete one to add another.",
        )

    with get_conn() as c:
        with c.cursor() as cur:
            try:
                cur.execute(
                    """INSERT INTO price_alerts (user_id, symbol, condition, target)
                       VALUES (%s, %s, %s, %s)
                       RETURNING id, symbol, condition, target, triggered, created_at""",
                    [user.id, payload.symbol, payload.condition, payload.target],
                )
                row = cur.fetchone()
            except Exception as e:
                if "unique" in str(e).lower():
                    raise HTTPException(
                        status_code=status.HTTP_409_CONFLICT,
                        detail="Identical alert already exists.",
                    )
                raise

    return {
        "id": row[0],
        "symbol": row[1],
        "condition": row[2],
        "target": float(row[3]),
        "triggered": row[4],
        "created_at": row[5].isoformat() if row[5] else None,
    }


@router.delete("/{alert_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_alert(alert_id: int, user: User = Depends(get_current_user)) -> None:
    """Delete a price alert (must belong to authenticated user)."""
    with get_conn() as c:
        with c.cursor() as cur:
            cur.execute(
                "SELECT id FROM price_alerts WHERE id = %s AND user_id = %s",
                [alert_id, user.id],
            )
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail="Alert not found")
            cur.execute(
                "DELETE FROM price_alerts WHERE id = %s AND user_id = %s",
                [alert_id, user.id],
            )
    return None

File: worker/src/routes/alerts.ts
// Price alert CRUD — port of api/routes/alerts.py (open to all signed-in users).

import { Hono } from "hono";
import { z } from "zod";
import { requireUser } from "../middleware";
import type { AppContext } from "../types";

const ALERT_LIMIT = 10;

const createSchema = z.object({
  symbol: z
    .string()
    .transform((s) => s.trim().toUpperCase())
    .pipe(z.string().regex(/^[A-Z][A-Z0-9.\-]{0,9}$/, "Invalid symbol format")),
  condition: z.enum(["above", "below"]),
  target: z.number().positive().transform((v) => Math.round(v * 10000) / 10000),
});

export const alertRoutes = new Hono<AppContext>();

alertRoutes.get("/", requireUser, async (c) => {
  const user = c.get("user")!;
  const rows = await c.env.DB.prepare(
    `SELECT id, symbol, condition, target, triggered, created_at
     FROM price_alerts WHERE user_id = ? ORDER BY created_at DESC`,
  )
    .bind(user.id)
    .all<{ id: number; symbol: string; condition: string; target: number; triggered: number; created_at: string }>();
  return c.json(
    (rows.results ?? []).map((r) => ({ ...r, triggered: !!r.triggered })),
  );
});

alertRoutes.post("/", requireUser, async (c) => {
  const parsed = createSchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) {
    return c.json({ detail: parsed.error.issues[0]?.message ?? "Invalid payload" }, 422);
  }
  const user = c.get("user")!;
  const { symbol, condition, target } = parsed.data;

  const activeCount = await c.env.DB.prepare(
    "SELECT COUNT(*) AS n FROM price_alerts WHERE user_id = ? AND triggered = 0",
  )
    .bind(user.id)
    .first<{ n: number }>();
  if ((activeCount?.n ?? 0) >= ALERT_LIMIT) {
    return c.json(
      { detail: `Alert limit reached (${ALERT_LIMIT} active alerts max). Delete one to add another.` },
      429,
    );
  }

  try {
    const row = await c.env.DB.prepare(
      `INSERT INTO price_alerts (user_id, symbol, condition, target)
       VALUES (?, ?, ?, ?)
       RETURNING id, symbol, condition, target, triggered, created_at`,
    )
      .bind(user.id, symbol, condition, target)
      .first<{ id: number; symbol: string; condition: string; target: number; triggered: number; created_at: string }>();
    return c.json({ ...row, triggered: !!row?.triggered }, 201);
  } catch (e) {
    if (String(e).toLowerCase().includes("unique")) {
      return c.json({ detail: "Identical alert already exists." }, 409);
    }
    throw e;
  }
});

alertRoutes.delete("/:id", requireUser, async (c) => {
  const id = Number(c.req.param("id"));
  if (!Number.isInteger(id)) return c.json({ detail: "Invalid id" }, 400);
  const user = c.get("user")!;
  const existing = await c.env.DB.prepare(
    "SELECT 1 FROM price_alerts WHERE id = ? AND user_id = ?",
  )
    .bind(id, user.id)
    .first();
  if (!existing) return c.json({ detail: "Alert not found" }, 404);
  await c.env.DB.prepare("DELETE FROM price_alerts WHERE id = ? AND user_id = ?")
    .bind(id, user.id)
    .run();
  return c.body(null, 204);
});
